File: java-core/src/main/java/software/amazon/lambda/durable/logging/DurableLogger.java
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
package software.amazon.lambda.durable.logging;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.slf4j.MDC;
import software.amazon.lambda.durable.DurableContext;
import software.amazon.lambda.durable.StepContext;
import software.amazon.lambda.durable.context.BaseContext;
import software.amazon.lambda.durable.model.SafeCloseable;

/**
 * Logger wrapper that adds durable execution context to log entries via MDC and optionally suppresses logs during
 * replay.
 */
public class DurableLogger {
    static final String MDC_DURABLE_EXECUTION_ARN = "durableExecutionArn";
    static final String MDC_EXECUTION_ARN = "executionArn";
    static final String MDC_REQUEST_ID = "requestId";
    static final String MDC_OPERATION_ID = "operationId";
    static final String MDC_CONTEXT_ID = "contextId";
    static final String MDC_OPERATION_NAME = "operationName";
    static final String MDC_CONTEXT_NAME = "contextName";
    static final String MDC_ATTEMPT = "attempt";

    public static final DurableLogger INSTANCE = new DurableLogger(LoggerFactory.getLogger(DurableLogger.class));
    private static final SafeCloseable AUTO_CLOSER = DurableLogger::detachContext;

    private final Logger delegate;

    /**
     * Creates a DurableLogger wrapping the given SLF4J logger with execution context MDC entries.
     *
     * @param delegate the SLF4J logger to wrap
     */
    public DurableLogger(Logger delegate) {
        this.delegate = delegate;
    }

    /** Returns the shared context-aware durable logger. */
    public static DurableLogger getLogger() {
        return INSTANCE;
    }

    /**
     * Returns a context-aware durable logger wrapping the given SLF4J logger.
     *
     * @param delegate the SLF4J logger to wrap
     */
    public static DurableLogger getLogger(Logger delegate) {
        return new DurableLogger(delegate);
    }

    public static SafeCloseable attachContext() {
        var context = BaseContext.getCurrentContext();
        if (context != null) {
            injectMdcProperties(context);
        }
        return AUTO_CLOSER;
    }

    public static void detachContext() {
        var context = BaseContext.getCurrentContext();
        if (context != null) {
            MDC.clear();
        }
    }

    private static void injectMdcProperties(BaseContext context) {
        var config = context.getDurableConfig().getLoggerConfig();

        // execution arn
        if (config.oldKeyNames()) {
            MDC.put(MDC_DURABLE_EXECUTION_ARN, context.getExecutionArn());
        } else {
            MDC.put(MDC_EXECUTION_ARN, context.getExecutionArn());
        }

        // lambda request id
        var requestId =
                context.getLambdaContext() != null ? context.getLambdaContext().getAwsRequestId() : null;
        if (requestId != null) {
            MDC.put(MDC_REQUEST_ID, requestId);
        }

        if (context instanceof DurableContext) {
            // context thread - context id and name
            if (context.getContextId() != null) {
                if (config.oldKeyNames()) {
                    MDC.put(MDC_CONTEXT_ID, context.getContextId());
                } else {
                    MDC.put(MDC_OPERATION_ID, context.getContextId());
                }
            }
            if (context.getContextName() != null) {
                if (config.oldKeyNames()) {
                    MDC.put(MDC_CONTEXT_NAME, context.getContextName());
                } else {
                    MDC.put(MDC_OPERATION_NAME, context.getContextName());
                }
            }
        } else if (context instanceof StepContext stepContext) {
            // In step context, context id is the operation id, context name is the operation name
            var operationId = context.getContextId();
            MDC.put(MDC_OPERATION_ID, operationId);
            if (context.getContextName() != null) {
                MDC.put(MDC_OPERATION_NAME, context.getContextName());
            }
            MDC.put(MDC_ATTEMPT, String.valueOf(stepContext.getAttempt()));
        }
    }

    public void trace(String format, Object... args) {
        log(() -> delegate.trace(format, args));
    }

    public void debug(String format, Object... args) {
        log(() -> delegate.debug(format, args));
    }

    public void info(String format, Object... args) {
        log(() -> delegate.info(format, args));
    }

    public void warn(String format, Object... args) {
        log(() -> delegate.warn(format, args));
    }

    public void error(String format, Object... args) {
        log(() -> delegate.error(format, args));
    }

    public void error(String message, Throwable t) {
        log(() -> delegate.error(message, t));
    }

    private boolean shouldSuppress(BaseContext context) {
        return context instanceof DurableContext durableContext
                && context.getDurableConfig().getLoggerConfig().suppressReplayLogs()
                && durableContext.isReplaying();
    }

    private void log(Runnable logAction) {
        var threadLocalContext = BaseContext.getCurrentContext();
        if (!shouldSuppress(threadLocalContext)) {
            logAction.run();
        }
    }
}

File: core/src/main/kotlin/io/github/zhongkechen/durable/DurableLogger.kt
package io.github.zhongkechen.durable

import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.slf4j.MDC

public class DurableLogger internal constructor(
    private val executionArn: String,
    private val operationId: String? = null,
    private val operationName: String? = null,
    private val attempt: Int? = null,
    private val delegate: Logger = LoggerFactory.getLogger("durable-execution"),
) {
    public fun trace(
        message: String,
        vararg arguments: Any?,
    ): Unit = withContext { delegate.trace(message, *arguments) }

    public fun debug(
        message: String,
        vararg arguments: Any?,
    ): Unit = withContext { delegate.debug(message, *arguments) }

    public fun info(
        message: String,
        vararg arguments: Any?,
    ): Unit = withContext { delegate.info(message, *arguments) }

    public fun warn(
        message: String,
        vararg arguments: Any?,
    ): Unit = withContext { delegate.warn(message, *arguments) }

    public fun error(
        message: String,
        throwable: Throwable? = null,
    ): Unit =
        withContext {
            if (throwable == null) delegate.error(message) else delegate.error(message, throwable)
        }

    private inline fun withContext(block: () -> Unit) {
        val previous =
            mapOf(
                "executionArn" to MDC.get("executionArn"),
                "operationId" to MDC.get("operationId"),
                "operationName" to MDC.get("operationName"),
                "attempt" to MDC.get("attempt"),
            )
        try {
            MDC.put("executionArn", executionArn)
            operationId?.let { MDC.put("operationId", it) }
            operationName?.let { MDC.put("operationName", it) }
            attempt?.let { MDC.put("attempt", it.toString()) }
            block()
        } finally {
            previous.forEach { (key, value) ->
                if (value == null) MDC.remove(key) else MDC.put(key, value)
            }
        }
    }
}
